File: examples/android/RunAnywhereAiNPU/app/src/main/java/com/runanywhere/sdk/runanywhereainpu/ui/screens/ModelsScreen.kt
package com.runanywhere.sdk.runanywhereainpu.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.downloadModel
import com.runanywhere.sdk.public.extensions.listModels
import com.runanywhere.sdk.public.extensions.loadModel
import com.runanywhere.sdk.public.types.RAModelInfo
import com.runanywhere.sdk.public.types.RAModelLoadRequest
import com.runanywhere.sdk.runanywhereainpu.ui.components.SectionCard
import com.runanywhere.sdk.runanywhereainpu.ui.theme.RaSuccess
import com.runanywhere.sdk.runanywhereainpu.ui.theme.Spacing
import kotlinx.coroutines.launch

@Composable
fun ModelsScreen() {
    val scope = rememberCoroutineScope()
    var models by remember { mutableStateOf<List<RAModelInfo>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }
    var status by remember { mutableStateOf<String?>(null) }
    val progress = remember { mutableStateMapOf<String, Float>() }

    suspend fun refresh() {
        loading = true; error = null
        try {
            val result = RunAnywhere.listModels()
            models = result.models?.models ?: emptyList()
        } catch (e: Exception) {
            error = e.message ?: "failed to list models"
        } finally {
            loading = false
        }
    }

    LaunchedEffect(Unit) { refresh() }

    Column(Modifier.fillMaxSize().padding(Spacing.md), verticalArrangement = Arrangement.spacedBy(Spacing.md)) {
        status?.let { Text(it, color = RaSuccess, style = MaterialTheme.typography.bodyMedium) }
        error?.let { Text(it, color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodyMedium) }

        when {
            loading -> Row(Modifier.fillMaxWidth().padding(Spacing.lg), horizontalArrangement = Arrangement.Center) {
                CircularProgressIndicator()
            }
            models.isEmpty() -> SectionCard {
                Text(
                    "No models in the registry. Models are added via the SDK catalog or " +
                        "registerModel(); QHexRT bundles are arch-specific (v79/v81).",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            else -> LazyColumn(verticalArrangement = Arrangement.spacedBy(Spacing.sm)) {
                items(models, key = { it.id }) { model ->
                    ModelRow(
                        model = model,
                        progress = progress[model.id],
                        onDownload = {
                            scope.launch {
                                progress[model.id] = 0f
                                try {
                                    RunAnywhere.downloadModel(model) { p ->
                                        progress[model.id] = p.stage_progress
                                    }
                                    progress.remove(model.id)
                                    status = "Downloaded ${model.name}"
                                    refresh()
                                } catch (e: Exception) {
                                    progress.remove(model.id)
                                    error = e.message ?: "download failed"
                                }
                            }
                        },
                        onLoad = {
                            scope.launch {
                                try {
                                    val r = RunAnywhere.loadModel(RAModelLoadRequest(model_id = model.id))
                                    status = if (r.success) "Loaded ${model.name}" else (r.error_message ?: "load failed")
                                } catch (e: Exception) {
                                    error = e.message ?: "load failed"
                                }
                            }
                        },
                    )
                }
            }
        }
    }
}

@Composable
private fun ModelRow(
    model: RAModelInfo,
    progress: Float?,
    onDownload: () -> Unit,
    onLoad: () -> Unit,
) {
    val downloaded = model.local_path.isNotBlank()
    SectionCard {
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(Spacing.sm),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(Spacing.xs)) {
                Text(
                    model.name.ifBlank { model.id },
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                )
                Row(horizontalArrangement = Arrangement.spacedBy(Spacing.xs)) {
                    Chip(
                        model.framework.name.removePrefix("INFERENCE_FRAMEWORK_").lowercase(),
                        MaterialTheme.colorScheme.primary,
                    )
                    if (model.download_size_bytes > 0) {
                        Chip(formatBytes(model.download_size_bytes), MaterialTheme.colorScheme.onSurfaceVariant)
                    }
                }
            }
            // One control that morphs by state: Download -> (progress) -> Use.
            when {
                progress != null -> Text(
                    "${(progress.coerceIn(0f, 1f) * 100).toInt()}%",
                    style = MaterialTheme.typography.labelLarge,
                    color = MaterialTheme.colorScheme.primary,
                )
                downloaded -> Button(onClick = onLoad) { Text("Use") }
                else -> Button(onClick = onDownload) { Text("Download") }
            }
        }
        if (progress != null) {
            LinearProgressIndicator(
                progress = { progress.coerceIn(0f, 1f) },
                modifier = Modifier.fillMaxWidth().padding(top = Spacing.sm),
            )
        }
    }
}

@Composable
private fun Chip(label: String, color: androidx.compose.ui.graphics.Color) {
    Text(
        label,
        style = MaterialTheme.typography.labelSmall,
        color = color,
        modifier = Modifier
            .background(color.copy(alpha = 0.12f), RoundedCornerShape(Spacing.xs))
            .padding(horizontal = 6.dp, vertical = 2.dp),
    )
}

private fun formatBytes(bytes: Long): String {
    if (bytes <= 0) return "—"
    val units = arrayOf("B", "KB", "MB", "GB")
    var v = bytes.toDouble()
    var i = 0
    while (v >= 1024 && i < units.lastIndex) {
        v /= 1024; i++
    }
    return if (i == 0) "$bytes B" else "%.1f %s".format(v, units[i])
}

File: examples/flutter/RunAnywhereAiNPU/lib/screens/models_screen.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:runanywhere/runanywhere.dart';
import 'package:runanywhere/runanywhere_protos.dart' as ra;

import '../theme.dart';
import '../widgets.dart';

class ModelsScreen extends StatefulWidget {
  const ModelsScreen({super.key});

  @override
  State<ModelsScreen> createState() => _ModelsScreenState();
}

class _ModelsScreenState extends State<ModelsScreen> {
  List<ra.ModelInfo> _models = [];
  bool _loading = true;
  String? _error;
  String? _status;
  final Map<String, double> _progress = {};

  @override
  void initState() {
    super.initState();
    _refresh();
  }

  Future<void> _refresh() async {
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      final models = await RunAnywhere.models.available();
      setState(() => _models = models);
    } catch (e) {
      setState(() => _error = '$e');
    } finally {
      setState(() => _loading = false);
    }
  }

  void _download(ra.ModelInfo model) {
    setState(() => _progress[model.id] = 0);
    RunAnywhere.downloads.start(model.id).listen(
      (p) {
        setState(() => _progress[model.id] = p.overallProgress);
        if (p.state == ra.DownloadState.DOWNLOAD_STATE_COMPLETED) {
          setState(() {
            _progress.remove(model.id);
            _status = 'Downloaded ${model.name}';
          });
          _refresh();
        } else if (p.state == ra.DownloadState.DOWNLOAD_STATE_FAILED) {
          setState(() {
            _progress.remove(model.id);
            _error = p.errorMessage.isEmpty ? 'Download failed' : p.errorMessage;
          });
        }
      },
      onError: (e) => setState(() {
        _progress.remove(model.id);
        _error = '$e';
      }),
    );
  }

  Future<void> _load(ra.ModelInfo model) async {
    try {
      await RunAnywhere.models.loadModel(model);
      setState(() => _status = 'Loaded ${model.name}');
    } catch (e) {
      setState(() => _error = '$e');
    }
  }

  @override
  Widget build(BuildContext context) {
    final cs = Theme.of(context).colorScheme;
    return Scaffold(
      appBar: AppBar(title: const Text('Models')),
      body: ResponsiveBody(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            if (_status != null) Text(_status!, style: TextStyle(color: raSuccess)),
            if (_error != null) Text(_error!, style: TextStyle(color: cs.error)),
            const SizedBox(height: 8),
            if (_loading)
              const Center(child: Padding(padding: EdgeInsets.all(24), child: CircularProgressIndicator()))
            else if (_models.isEmpty)
              SectionCard(
                child: Text(
                  'No models in the registry. Models are added via the SDK catalog '
                  'or register(); QHexRT bundles are arch-specific (v79/v81).',
                  style: Theme.of(context).textTheme.bodyMedium?.copyWith(color: cs.onSurfaceVariant),
                ),
              )
            else
              for (final model in _models) ...[
                _ModelCard(
                  model: model,
                  progress: _progress[model.id],
                  onDownload: () => _download(model),
                  onLoad: () => _load(model),
                ),
                const SizedBox(height: 8),
              ],
          ],
        ),
      ),
    );
  }
}

class _ModelCard extends StatelessWidget {
  const _ModelCard({required this.model, this.progress, required this.onDownload, required this.onLoad});

  final ra.ModelInfo model;
  final double? progress;
  final VoidCallback onDownload;
  final VoidCallback onLoad;

  @override
  Widget build(BuildContext context) {
    final cs = Theme.of(context).colorScheme;
    final downloaded = model.localPath.isNotEmpty || model.isDownloaded;
    final fw = model.framework.name.replaceFirst('INFERENCE_FRAMEWORK_', '').toLowerCase();
    return SectionCard(
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(model.name.isEmpty ? model.id : model.name,
                    style: Theme.of(context).textTheme.titleMedium),
                const SizedBox(height: 4),
                Row(children: [
                  _chip(context, fw, cs.primary),
                  const SizedBox(width: 4),
                  if (model.downloadSizeBytes.toInt() > 0)
                    _chip(context, _formatBytes(model.downloadSizeBytes.toInt()), cs.onSurfaceVariant),
                ]),
              ],
            ),
          ),
          const SizedBox(width: 8),
          if (progress != null)
            Text('${(progress!.clamp(0, 1) * 100).toInt()}%',
                style: metricTextStyle.copyWith(color: cs.primary))
          else if (downloaded)
            FilledButton(onPressed: onLoad, child: const Text('Use'))
          else
            FilledButton(onPressed: onDownload, child: const Text('Download')),
        ],
      ),
    );
  }

  Widget _chip(BuildContext context, String label, Color color) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
      decoration: BoxDecoration(
        color: color.withValues(alpha: 0.12),
        borderRadius: BorderRadius.circular(4),
      ),
      child: Text(label, style: Theme.of(context).textTheme.labelSmall?.copyWith(color: color)),
    );
  }

  String _formatBytes(int bytes) {
    if (bytes <= 0) return '—';
    const units = ['B', 'KB', 'MB', 'GB'];
    var v = bytes.toDouble();
    var i = 0;
    while (v >= 1024 && i < units.length - 1) {
      v /= 1024;
      i++;
    }
    return i == 0 ? '$bytes B' : '${v.toStringAsFixed(1)} ${units[i]}';
  }
}
